internal/shim-sev/src/hostcall.rs

```diff
@@ -64,175 +89,509 @@ fn return_empty_option<'a>(_i: usize) -> Option<&'a mut Block> {
+    // Safety: Split up mutable references to `SALLYPORT`, which can only be handed out
+    // via the `RwLocked` `HostCallAllocator`
+    let block_mut = &mut unsafe { &mut *SALLYPORT.get() }.0;
 
-    assert!(nr_syscall_blocks <= 512);
+    let hostcall_iter = hostcall_allocator.0.iter_mut();
 
-    let mut hostcall_allocator = HostCallAllocator(array_const_fn_init![return_empty_option; 512]);
-
-    for i in 0..nr_syscall_blocks {
-        (hostcall_allocator.0)[i].replace(unsafe { &mut *block_mut.add(i) });
+    for (store, chunk) in hostcall_iter.zip(block_mut.chunks_mut(BLOCK_SIZE_USIZE)) {
+        store.replace(chunk);
     }
 
     RwLocked::<HostCallAllocator>::new(hostcall_allocator)
 });
 
 /// Allocator for all `sallyport::Block`
-pub struct HostCallAllocator([Option<&'static mut Block>; MAX_BLOCK_NR]);
+#[derive(Default)]
+pub struct HostCallAllocator([Option<&'static mut [usize]>; NUM_BLOCKS]);
 
 impl RwLocked<HostCallAllocator> {
     /// Try to allocate a `HostCall` object to use a `sallyport::Block`
-    pub fn try_alloc(&self) -> Option<HostCall> {
+    pub fn try_alloc(&self) -> Option<BlockGuard> {
         let mut this = self.write();
         this.0
             .iter_mut()
             .enumerate()
             .find(|(_i, x)| x.is_some())
-            .map(|(i, ele)| HostCall {
+            .map(|(i, ele)| BlockGuard {
                 block_index: i as _,
                 block: ele.take(),
             })
     }
 }
 
 /// Communication with the Host
-pub struct HostCall {
+pub struct BlockGuard {
     block_index: u16,
-    block: Option<&'static mut Block>,
+    block: Option<&'static mut [usize]>,
 }
 
-impl Drop for HostCall {
+impl Drop for BlockGuard {
     fn drop(&mut self) {
         HOST_CALL_ALLOC.write().0[self.block_index as usize] = self.block.take();
     }
 }
 
-impl HostCall {
+/// The syscall Handler
+pub struct HostCall<'a> {
+    block_guard: BlockGuard,
+    tls: &'a mut ThreadLocalStorage,
+}
+
+impl<'a> HostCall<'a> {
+    /// Try to get a new instance
+    pub fn try_new(tls: &'a mut ThreadLocalStorage) -> Option<HostCall<'a>> {
+        let bg = HOST_CALL_ALLOC.try_alloc()?;
+
+        Some(Self {
+            block_guard: bg,
+            tls,
+        })
+    }
+
+    /// get an SNP attestation report
+    ///
+    /// See https://github.com/enarx/enarx/issues/966
+    pub fn get_attestation(
+        &mut self,
+        nonce: usize,
+        nonce_len: usize,
+        buf: usize,
+        buf_len: usize,
+    ) -> Result<[usize; 2], libc::c_int> {
+        if !snp_active() {
+            return Ok([0, 0]);
+        }
+
+        if buf == 0 {
+            return Ok([SNP_ATTESTATION_LEN_MAX, TECH]);
+        }
+
+        if buf_len > isize::MAX as usize {
+            return Err(libc::EINVAL);
+        }
+
+        if buf_len < SNP_ATTESTATION_LEN_MAX {
+            return Err(libc::EMSGSIZE);
+        }
+
+        if nonce_len != 64 {
+            return Err(libc::EINVAL);
+        }
```

Comment:
How about we merge all of these into one `match` statement with guards?

Reply:
doesn't look good